src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h

```diff
@@ -0,0 +1,366 @@
+        use_gpu_feature_extraction(true) {}
+  void Register(OptionsItf *po) {
+    po->Register("max-batch-size", &max_batch_size,
+                 "The maximum execution batch size. "
+                 "Larger = Better throughput slower latency.");
+    po->Register("num-channels", &num_channels,
+                 "The number of parallel audio channels. This is the maximum "
+                 "number of parallel audio channels supported by the pipeline"
+                 ". This should be larger "
+                 "than max_batch_size.");
+    po->Register("cuda-worker-threads", &num_worker_threads,
+                 "(optional) The total number of CPU threads launched to "
+                 "process CPU tasks. -1 = use std::hardware_concurrency()");
+    po->Register("determinize-lattice", &determinize_lattice,
+                 "Determinize the lattice before output.");
+    po->Register("cuda-decoder-copy-threads", &num_decoder_copy_threads,
+                 "Advanced - Number of worker threads used in the "
+                 "decoder for "
+                 "the host to host copies.");
+    po->Register("gpu-feature-extract", &use_gpu_feature_extraction,
+                 "Use GPU feature extraction");
+
+    feature_opts.Register(po);
+    decoder_opts.Register(po);
+    det_opts.Register(po);
+    compute_opts.Register(po);
+  }
+  int max_batch_size;
+  int num_channels;
+  int num_worker_threads;
+  bool determinize_lattice;
+  int num_decoder_copy_threads;
+  bool use_gpu_feature_extraction;
+
+  OnlineNnet2FeaturePipelineConfig feature_opts;
+  CudaDecoderConfig decoder_opts;
+  fst::DeterminizeLatticePhonePrunedOptions det_opts;
+  nnet3::NnetSimpleComputationOptions compute_opts;
+
+  void CheckAndFixConfigs() {
+    KALDI_ASSERT(max_batch_size > 0);
+    // Lower bound on nchannels.
+    // Using strictly more than max_batch_size because channels are still used
+    // when the lattice postprocessing is running. We still want to run full
+    // max_batch_size batches in the meantime
+    int min_nchannels =
+        max_batch_size * KALDI_CUDA_DECODER_MIN_NCHANNELS_FACTOR;
+    num_channels = std::max(num_channels, min_nchannels);
+
+    // If not set use number of physical threads
+    num_worker_threads = (num_worker_threads > 0)
+                             ? num_worker_threads
+                             : std::thread::hardware_concurrency();
+  }
+};
+
+class BatchedThreadedNnet3CudaOnlinePipeline {
+ public:
+  using CorrelationID = uint64_t;
+  BatchedThreadedNnet3CudaOnlinePipeline(
+      const BatchedThreadedNnet3CudaOnlinePipelineConfig &config,
+      const fst::Fst<fst::StdArc> &decode_fst,
+      const nnet3::AmNnetSimple &am_nnet, const TransitionModel &trans_model)
+      : config_(config),
+        max_batch_size_(config.max_batch_size),
+        trans_model_(&trans_model),
+        am_nnet_(&am_nnet),
+        word_syms_(NULL) {
+    config_.compute_opts.CheckAndFixConfigs(am_nnet_->GetNnet().Modulus());
```

Comment:
Is this missing a call to `config_.CheckAndFixConfigs()`? Without it, the default value of --cuda-worker-threads (-1) will lead to an error.

Reply:
Thanks for catching this! I'll update it